src/random_ray/random_ray.cpp

```diff
@@ -0,0 +1,284 @@
+    }
+    distance_travelled_ += distance;
+    attenuate_flux(distance, true);
+  }
+
+  // Check for end of inactive region (dead zone)
+  if (!is_active_) {
+    if (distance_travelled_ + distance >=
+        settings::random_ray_distance_inactive) {
+      is_active_ = true;
+      double distance_dead =
+        settings::random_ray_distance_inactive - distance_travelled_;
+      attenuate_flux(distance_dead, false);
+
+      double distance_alive = distance - distance_dead;
+
+      // Ensure we haven't travelled past the active phase as well
+      if (distance_alive > settings::random_ray_distance_active) {
+        distance_alive = settings::random_ray_distance_active;
+        wgt() = 0.0;
+      }
+      attenuate_flux(distance_alive, true);
+
+      distance_travelled_ = distance_alive;
+    } else {
+      distance_travelled_ += distance;
+      attenuate_flux(distance, false);
+    }
+  }
+
+  // Advance particle
+  for (int j = 0; j < n_coord(); ++j) {
+    coord(j).r += distance * coord(j).u;
+  }
+}
+
+// This function forms the inner loop of the random ray transport process.
+// It is responsible for several tasks. Based on the incoming angular flux
+// of the ray and the source term in the region, the outgoing angular flux
+// is computed. The delta psi between the incoming and outgoing fluxes is
+// contributed to the estimate of the total scalar flux in the source region.
+// Additionally, the contribution of the ray path to the stochastically
+// estimated volume is also kept track of. All tasks involving writing
+// to the data for the source region are done with a lock over the entire
+// source region.  Locks are used instead of atomics as all energy groups
+// must be written, such that locking once is typically much more efficient
+// than use of many atomic operations corresponding to each energy group
+// individually (at least on CPU). Several other bookeeping tasks are also
+// performed when inside the lock.
+void RandomRay::attenuate_flux(double distance, bool is_active)
+{
+  int negroups = data::mg.num_energy_groups_;
+
+  // The number of geometric intersections is counted for reporting purposes
+  n_event()++;
+
+  // Determine source region index etc.
+  int i_cell = lowest_coord().cell;
+
+  // The source region is the spatial region index
+  int64_t source_region =
+    random_ray::source_region_offsets[i_cell] + cell_instance();
+
+  // The source element is the energy-specific region index
+  int64_t source_element = source_region * negroups;
+  int material = this->material();
+
+  // Temperature and angle indices, if using multiple temperature
+  // data sets and/or anisotropic data sets.
+  // TODO: Currently assumes we are only using single temp/single
+  // angle data.
+  const int t = 0;
+  const int a = 0;
+
+  // MOC incoming flux attenuation + source contribution/attenuation equation
+  for (int e = 0; e < negroups; e++) {
+    float Sigma_t = data::mg.macro_xs_[material].get_xs(
```

Comment:
`get_xs` is probably not going to get inlined here due to being in another translation unit. For a performance-critical loop like this, it might give better results if `get_xs` could somehow be replaced with the raw access into an array. But that's just a hunch. Then it _might_ automatically vectorize. That'd be epic.

Reply:
For sure, the current XS access patten is not great. It hurts us even more in the iteration source update function, as that really hammers on the scattering matrix in a bad pattern. In another implementation, I actually copy the data out into a bespoke format for random ray that is faster for accessing, but was planning to leave that as a follow up PR to keep things relatively simple for this one.

Reply:
Makes sense!